### ISSUE TITLE
Reduce server costs

### ISSUE THREAD

**Issue by simonhac** [opennem/opennem#159]
We're currently paying ~$3k/month in server costs — can we get this down without spending more than it's worth?

<img width="983" alt="image" src="https://user-images.githubusercontent.com/20500649/202385262-92d19bbc-b25b-42ab-bbc5-5454dfe64c3f.png">

**Comment by nc9** [comment]
ACK

**Comment by nc9** [comment]
Go through CPU / disk logs on database server and figure out ultimate instance type

**Comment by nc9** [comment]
Next release will also build ARM images here with a view of migrating to gravitron servers eventually which are ~30-40% more cost effective

**Comment by simonhac** [comment]
i suspect the server is 1-2 orders of magnitude too busy. 

it current takes more than 1.5-5 minutes to import  and process 426 records every 5 minutes. once the data has landed (ie. excluding network time), i can’t see why it needs to take more than a second or two of CPU time to insert records, update summaries, compute flows and spit out json files.

**Comment by nc9** [comment]
Three pronged approach @gerardmason 

- [ ] switch prod docker stack to Fargate (`infrav2`)
- [ ] commit database to 12-24 month buy
- [ ] profile database CPU and memory usage
- [ ] get Gerard onto ssh / tailscale network

**Comment by nc9** [comment]
#220 is related

i'm going to reduce the disk size by ~50%

**Comment by nc9** [comment]
primary `db` disk reduced from `2.2TB` to `800MB` and EBS volume type changed:

```
root@opennem03:~# df -h /data/volume2/
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme1n1    800G  518G  283G  65% /data/volume2
```

`prod` disk reduced from `220GB` to `60GB` 

we should see this propagate in cost estimator today

**Comment by simonhac** [comment]
@nc9 great progress… keen to see this run for a few days and see where costs stabilise.

before we prepurchase 12–24 month of database, i think we should get more confidence around the specs of the machine needed.

i'd like us to have a good idea of what the database is currently doing, so we can consider the benefit/cost of optimisations.

**Comment by simonhac** [comment]
@nc9 i know it's a minuscule expense, but why are we paying monthly for RDS?

<img width="517" alt="image" src="https://user-images.githubusercontent.com/20500649/225290435-21b0f88b-3b97-4962-b3f9-277baee9798e.png">

**Comment by nc9** [comment]
> i think we should get more confidence around the specs of the machine needed.

100% agree - need to do a full dive into all the different machine types and further profile our usage.

> i know it's a minuscule expense, but why are we paying monthly for RDS?

<img width="969" alt="image" src="https://user-images.githubusercontent.com/39009487/225449779-cccb77f3-947e-46f0-8f39-1ffeacb87e43.png">

deleted now

**Comment by nc9** [comment]
![image](https://user-images.githubusercontent.com/39009487/225816720-76c18161-eac6-4a1a-b8ed-9f00bbd30fa0.png)

![image](https://user-images.githubusercontent.com/39009487/225816762-431112f0-3f29-4ed5-ac7f-b42c57d8d0e6.png)

settled at `$19.50` per month. i'll reduce billing alerts to $20 per month

fargate migration should reduce this by another $3-4~ per day

**Comment by simonhac** [comment]
$19.50 per day, right?

**Comment by nc9** [comment]
Yep! alert set at $20

**Comment by nc9** [comment]
@simonhac want to close this off for now and we'll address `infraV3` in another ticket? /cc @gmmspi

**Comment by simonhac** [comment]
i see this as primarily a coding and db indexing issue, with infra being a secondary concern that cannot be solved until the primary issue is addressed.

i believe we are doing too much too often. i don't want to close this issue until both of these tasks are well under 5s, preferably 1s (from when the the CSV is fetched).

<img width="667" alt="image" src="https://user-images.githubusercontent.com/20500649/226252466-337f13da-697e-4b09-840d-581aaab1d6c1.png">

**Comment by nc9** [comment]
#221  linked

**Comment by gmmspi** [comment]
I don't think the processing times will affect server cost, the instances are always-on.

**Comment by simonhac** [comment]
if the database could be made 1/10th as busy, couldn't we significantly downsize the specs of the DB server? run everything on the same relatively small machine?

**Comment by nc9** [comment]
current DB only has 16GB of RAM and it's swapping on overnights already. the key will be to remove the persistent `prod` server that is running the docker stack and migrate that to fargate

lets see what these profiles look like with queries optimized in `3.14`

**Comment by gmmspi** [comment]
Maybe Simon has a point there, if it's swapping on those overnights we can target that for optimisation.

**Comment by simonhac** [comment]
we append 427 records every 5 minutes from the NEM to a time series, far fewer from WEM, APVI and BOM — and maintain a couple of hundred hierarchical summaries on a daily and monthly basis.

that's a tiny tiny trickle of data that should be able to be managed by a pretty low-rent database — if we're being efficient.

**Comment by simonhac** [comment]
the overnight data is currently taking 3.5 hours to process… if the database is working as expected, should this task take at most a few minutes?
![image](https://user-images.githubusercontent.com/20500649/226497155-c37e2289-65ba-4397-8615-584cad0eefb5.png)

**Comment by gmmspi** [comment]
Hey @nc9 is there logging in AWS for memory usage? I'd like to monitor how the db server is utilising it's resources. I think these nightlies need to be optimised before we can close this issue - if the monitoring shows an improvement after optimisation, we may be able to downsize the instance.

**Comment by simonhac** [comment]
half what we spent in november, but still a way to go:
<img width="965" alt="image" src="https://user-images.githubusercontent.com/20500649/230020751-9369450e-e1ce-41ed-a4ae-d3c41a6e6fc2.png">
